elastic/beats
# Enrich k8s state service events & add selector meta

## What does this PR do?
1. adds metadata enricher on `state_service` metrics so as to enrich metrics with Service objects metadata.
2. make use of `ServiceMetadataGenerator` to enrich Service related events and not the generic `ResourceMetadataGenerator`
3. Improves `ServiceMetadataGenerator` so as to include `selector` information (if exist) for a Service object.

Note that with this implementation `selector`'s key:val maps are added similarly with `labels` like `kubernetes.service.selectors.*` but we can add them differently if Metrics UI requires that, @sorantis just let me know what you think.

## Why is it important?
So as to properly enrich `state_service` events with Service metadata and include `selector` information in these metadata.

Example:
```
apiVersion: v1
kind: Service
metadata:
  name: my-service
spec:
  selector:
    app: MyApp
  ports:
    - protocol: TCP
      port: 80
      targetPort: 9376
```

This service will be enriched with `kubernetes.service.selectors.app: MyApp` which will give us to the option to correlate the service with all the pods that are selected/exposed by this service and are labeled with `kubernetes.labels.app: MyApp`.

## Checklist




## How to test this PR locally
1. Deploy Metricbeat on k8s to collect metrics from `kube_state_metrics` and specifically using `state_service` metricset.
2. Ensure that metadata are included (ie  labels) as well as selector information.

## Related issues
- Closes https://github.com/elastic/beats/issues/22444
- Relates to https://github.com/elastic/integrations/issues/511
